core/consensus/babe/impl/babe_synchronizer_impl.hpp

```diff
@@ -3,62 +3,100 @@
 
-#include "application/app_configuration.hpp"
-#include "log/logger.hpp"
-#include "network/types/sync_clients_set.hpp"
+#include <queue>
+
+#include <libp2p/basic/scheduler.hpp>
+
+#include "application/app_state_manager.hpp"
+#include "consensus/babe/impl/block_executor.hpp"
 #include "network/router.hpp"
 
 namespace kagome::consensus {
 
-  /**
-   * Implementation of babe synchronizer that requests blocks from provided
-   * peers
-   */
   class BabeSynchronizerImpl
       : public BabeSynchronizer,
         public std::enable_shared_from_this<BabeSynchronizerImpl> {
    public:
-    ~BabeSynchronizerImpl() override = default;
+    enum class Error {
+      SHUTTING_DOWN = 1,
+      EMPTY_RESPONSE,
+      WRONG_CONTENT,
+      DISCARDED_BLOCK,
+      WRONG_ORDER,
+      INVALID_HASH,
+      ALREADY_IN_QUEUE,
+      PEER_BUSY
+    };
 
     BabeSynchronizerImpl(
-        std::shared_ptr<network::SyncClientsSet> sync_clients,
-        const application::AppConfiguration &app_configuration,
-        std::shared_ptr<network::Router> router);
+        std::shared_ptr<application::AppStateManager> app_state_manager,
+        std::shared_ptr<blockchain::BlockTree> block_tree,
+        std::shared_ptr<BlockExecutor> block_executor,
+        std::shared_ptr<network::Router> router,
+        std::shared_ptr<libp2p::basic::Scheduler> scheduler,
+        std::shared_ptr<crypto::Hasher> hasher);
+
+    void enqueue(const primitives::BlockInfo &block_info,
+                 const libp2p::peer::PeerId &peer_id,
+                 SyncResultHandler &&handler) override;
 
-    void request(const primitives::BlockId &from,
-                 const primitives::BlockHash &to,
+    void enqueue(const primitives::BlockHeader &header,
                  const libp2p::peer::PeerId &peer_id,
-                 const BlocksHandler &block_list_handler) override;
+                 SyncResultHandler &&handler) override;
 
    private:
-    /**
-     * Select next client to be polled
-     * @param polled_clients clients that we already polled
-     * @return next clint to be polled
-     */
-    std::shared_ptr<network::SyncProtocolClient> selectNextClient(
-        std::unordered_set<std::shared_ptr<network::SyncProtocolClient>>
-            &polled_clients) const;
-    /**
-     * Request blocks from provided peers
-     * @param response_res block request message
-     * @param polled_clients peers that were already requested
-     * @param requested_blocks_handler handler of received blocks
-     */
-    void pollClients(network::BlocksRequest response_res,
-                     const libp2p::peer::PeerId &peer_id,
-                     const BlocksHandler &requested_blocks_handler) const;
-
-    std::shared_ptr<network::SyncClientsSet> sync_clients_;
-    log::Logger logger_;
-    const application::AppConfiguration &app_configuration_;
+    bool isInQueue(const primitives::BlockHash &hash) const;
+
+    void findCommonBlock(const libp2p::peer::PeerId &peer_id,
```

Comment:
Would really appreciate a short description of the algorithm of this method.

Reply:
Commented